components/sdmmc/sdmmc_cmd.c

```diff
@@ -263,61 +279,214 @@ esp_err_t sdmmc_card_init(const sdmmc_host_t* config, sdmmc_card_t* card)
+					err = (*config->set_card_clk)(config->slot, SDMMC_FREQ_HIGHSPEED);
+					if (err != ESP_OK) {
+						ESP_LOGE(TAG, "failed to switch peripheral to HS bus mode");
+						free(ext_csd);
+						return err;
+					}
+				} else if (config->max_freq_khz >= SDMMC_FREQ_DEFAULT &&
+						speed >= SDMMC_FREQ_DEFAULT) {
+					ESP_LOGD(TAG, "switching to DS bus mode");
+					err = (*config->set_card_clk)(config->slot, SDMMC_FREQ_DEFAULT);
+					if (err != ESP_OK) {
+						ESP_LOGE(TAG, "failed to switch peripheral to HS bus mode");
+						free(ext_csd);
+						return err;
+					}
+				}
+
+				if (timing != SDMMC_TIMING_LEGACY) {
+					/* read EXT_CSD again */
+					err = sdmmc_mem_send_cxd_data(card,
+							MMC_SEND_EXT_CSD, ext_csd, EXT_CSD_SIZE);
+					if (err != ESP_OK) {
+						ESP_LOGE(TAG, "%s: can't re-read EXT_CSD\n", __func__);
+						free(ext_csd);
+						return err;
+					}
+					if (ext_csd[EXT_CSD_HS_TIMING] != EXT_CSD_HS_TIMING_HS) {
+						ESP_LOGE(TAG, "%s, HS_TIMING set failed\n", __func__);
+						free(ext_csd);
+						return ESP_ERR_INVALID_RESPONSE;
+					}
+				}
+
+				if (card->host.flags & SDMMC_HOST_FLAG_8BIT) {
+					width = 8;
+					value = EXT_CSD_BUS_WIDTH_8;
+					powerclass = ext_csd[(speed > 26000) ? EXT_CSD_PWR_CL_52_360 : EXT_CSD_PWR_CL_26_360] >> 4;
+				} else if (card->host.flags & SDMMC_HOST_FLAG_4BIT) {
+					width = 4;
+					value = EXT_CSD_BUS_WIDTH_4;
+					powerclass = ext_csd[(speed > 26000) ? EXT_CSD_PWR_CL_52_360 : EXT_CSD_PWR_CL_26_360] & 0x0f;
+				} else {
+					width = 1;
+					value = EXT_CSD_BUS_WIDTH_1;
+					powerclass = 0; //card must be able to do full rate at powerclass 0 in 1-bit mode
+				}
+				if (powerclass != 0) {
+					err = sdmmc_mmc_switch(card, EXT_CSD_CMD_SET_NORMAL,
+							EXT_CSD_POWER_CLASS, powerclass);
+					if (err != ESP_OK) {
+						ESP_LOGE(TAG, "%s: can't change power class"
+									" (%d bit)\n", __func__, powerclass);
+						free(ext_csd);			
+						return err;
+					}
+				}
+				if (width != 1) {
+					err = sdmmc_mmc_switch(card, EXT_CSD_CMD_SET_NORMAL,
+							EXT_CSD_BUS_WIDTH, value);
+					if (err == ESP_OK) {
+						err = (*config->set_bus_width)(config->slot, width);
+						if (err != ESP_OK) {
+							ESP_LOGE(TAG, "slot->set_bus_width failed");
+							free(ext_csd);
+							return err;
+						}
+					} else {
+						ESP_LOGE(TAG, "%s: can't change bus width"
+									" (%d bit)\n", __func__, width);
+						free(ext_csd);			
+						return err;
+					}
+
+					/* XXXX: need bus test? (using by CMD14 & CMD19) */
+					ets_delay_us(10000);
+				}
+
+				sectors = (ext_csd[EXT_CSD_SEC_COUNT + 0] << 0)
+					| (ext_csd[EXT_CSD_SEC_COUNT + 1] << 8)  
+					| (ext_csd[EXT_CSD_SEC_COUNT + 2] << 16) 
+					| (ext_csd[EXT_CSD_SEC_COUNT + 3] << 24);
+
+				if (sectors > (2u * 1024 * 1024 * 1024) / 512) {
+					//card->flags |= SFF_SDHC;
+					card->csd.capacity = sectors;
+				}
+				
+				free(ext_csd); //done with ext_csd
+			}		
+	
+		}else{  //SD CARD
```

Comment:
Please add spaces between keywords and braces

Reply:
done